File: server/data_loaders/files.py
import os
import markdown
import PyPDF2
import pytesseract
from PIL import Image
import io
import hashlib
from server.embeddings.embed import get_embedding, chunk_content
from server.database.queries import create_file, get_file_by_content_hash
from server.constants import folder_path
import traceback

def get_file_contents(file_path: str) -> str:
    _, ext = os.path.splitext(file_path)
    
    if ext.lower() in ['.md', '.txt']:
        with open(file_path, 'r', encoding='utf-8') as file:
            content = file.read()
            if ext.lower() == '.md':
                content = markdown.markdown(content)
            return content
    elif ext.lower() == '.pdf':
        return extract_text_from_pdf(file_path)
    else:
        raise ValueError(f"Unsupported file type: {ext}")

def extract_text_from_pdf(file_path: str) -> str:
    text = ""
    with open(file_path, 'rb') as file:
        reader = PyPDF2.PdfReader(file)
        for page in reader.pages:
            text += page.extract_text()
    
    if not text.strip():  # If no text was extracted, use OCR
        return ocr_pdf(file_path)
    
    # Remove NUL characters
    return text.replace('\x00', '')

def ocr_pdf(file_path: str) -> str:
    text = ""
    with open(file_path, 'rb') as file:
        pdf = PyPDF2.PdfReader(file)
        for page in range(len(pdf.pages)):
            page_obj = pdf.pages[page]
            for image in page_obj.images:
                img = Image.open(io.BytesIO(image.data))
                text += pytesseract.image_to_string(img)
    
    # Remove NUL characters
    return text.replace('\x00', '')

def get_content_hash(content: str) -> str:
    return hashlib.sha256(content.encode()).hexdigest()

def process_files(folder_path: str):
    for root, _, files in os.walk(folder_path):
        for file in files:
            file_path = os.path.join(root, file)
            _, ext = os.path.splitext(file)
            
            if ext.lower() in ['.md', '.txt', '.pdf']:
                try:
                    content = get_file_contents(file_path)
                    chunks = chunk_content(content)
                    
                    for i, chunk in enumerate(chunks):
                        content_hash = get_content_hash(chunk)
                        existing_file = get_file_by_content_hash(content_hash)
                        if existing_file:
                            continue

                        embedding = get_embedding(chunk)
                        
                        file_data = {
                            "name": f"{file}_chunk_{i+1}",
                            "path": os.path.abspath(file_path),  # Use absolute path
                            "content": chunk,
                            "content_hash": content_hash,
                            "embedding": embedding
                        }
                        
                        stored_file = create_file(file_data)
                        print(f"Stored file chunk: {stored_file.id} - {stored_file.name}")
                except Exception as e:
                    print(f"Error processing file {file_path}: {e}")
                    print(traceback.format_exc())  # This will print the full stack trace

if __name__ == "__main__":
    folder_path = folder_path
    process_files(folder_path)

File: server/src/services/dataLoaders/files.ts
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { marked } from 'marked';
import { PDFDocument } from 'pdf-lib';
import { createWorker } from 'tesseract.js';
import { container } from '../../utils/container';
import { PrismaClient } from '@prisma/client';
import { getEmbedding, chunkContent } from '../embeddings/embed';
import { createFile } from '../../dao/fileDao';
import { createEmbedding } from '../../dao/embeddingDao';
import { fromPath } from 'pdf2pic';

const FOLDER_PATH = process.env.FOLDER_PATH || 'jeff_storage';

async function getFileContents(filePath: string): Promise<string> {
  const ext = path.extname(filePath).toLowerCase();

  if (ext === '.md' || ext === '.txt') {
    const content = fs.readFileSync(filePath, 'utf-8');
    return ext === '.md' ? marked(content) : content;
  } else if (ext === '.pdf') {
    return ocrPdf(filePath);
  } else {
    throw new Error(`Unsupported file type: ${ext}`);
  }
}

async function ocrPdf(filePath: string): Promise<string> {
  const worker = await createWorker();
  let text = '';

  const options = {
    density: 300,
    saveFilename: "temp_page",
    savePath: "./temp",
    format: "png",
    width: 2480,
    height: 3508
  };

  const convert = fromPath(filePath, options);
  const pdfDoc = await PDFDocument.load(fs.readFileSync(filePath));

  for (let i = 0; i < pdfDoc.getPageCount(); i++) {
    const result = await convert(i + 1);
    const base64 = (result as any).base64;
    const { data: { text: pageText } } = await worker.recognize(`data:image/png;base64,${base64}`);
    text += pageText + '\n';
  }

  await worker.terminate();
  return text.replace(/\0/g, '');
}

function getContentHash(content: string): string {
  return crypto.createHash('sha256').update(content).digest('hex');
}

export async function processFiles(folderPath: string): Promise<void> {
  const prisma = container.get(PrismaClient);

  const processFile = async (filePath: string) => {
    const ext = path.extname(filePath).toLowerCase();

    if (['.md', '.txt', '.pdf'].includes(ext)) {
      try {
        const content = await getFileContents(filePath);
        const chunks = chunkContent(content);

        for (let i = 0; i < chunks.length; i++) {
          const chunk = chunks[i];
          const contentHash = getContentHash(chunk);

          const existingFile = await prisma.file.findFirst({
            where: { contentHash }
          });

          if (existingFile) continue;

          const embedding = await getEmbedding(chunk);

          const fileData = {
            name: `${path.basename(filePath)}_chunk_${i + 1}`,
            path: filePath,
            content: chunk,
            contentHash: contentHash,
          };

          const storedFile = await createFile(fileData);
          await createEmbedding({
            embedding,
            contentType: 'FILE',
            fileId: storedFile.id,
          });

          console.log(`Stored file chunk: ${storedFile.id} - ${storedFile.name}`);
        }
      } catch (error) {
        console.error(`Error processing file ${filePath}:`, error);
      }
    }
  };

  const processDirectory = async (dir: string) => {
    const entries = fs.readdirSync(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        await processDirectory(fullPath);
      } else {
        await processFile(fullPath);
      }
    }
  };

  await processDirectory(folderPath);
}

if (require.main === module) {
  processFiles(FOLDER_PATH).catch(console.error);
}
